Detectors/TPC/calibration/src/CalculatedEdx.cxx

```diff
@@ -0,0 +1,484 @@
+                 << "sectorVector=" << sectorVector
+                 << "topologyCorrVector=" << topologyCorrVector
+                 << "topologyCorrTotVector=" << topologyCorrTotVector
+                 << "topologyCorrMaxVector=" << topologyCorrMaxVector
+                 << "gainVector=" << gainVector
+                 << "gainResidualVector=" << gainResidualVector
+                 << "residualCorrTotVector=" << residualCorrTotVector
+                 << "residualCorrMaxVector=" << residualCorrMaxVector
+                 << "xPositionVector=" << xPositionVector
+                 << "localYVector=" << localYVector
+                 << "offsPadVector=" << offsPadVector
+                 << "trackVector=" << trackVector
+                 << "clVector=" << clVector
+                 << "chargeTotVector=" << chargeTotVector
+                 << "chargeMaxVector=" << chargeMaxVector
+                 << "minChargeTot=" << minChargeTot
+                 << "minChargeMax=" << minChargeMax
+                 << "output=" << output
+                 << "\n";
+  }
+}
+
+float CalculatedEdx::getTruncMean(std::vector<float>& charge, float low, float high)
+{
+  // sort the charge vector
+  std::sort(charge.begin(), charge.end());
+
+  // calculate truncated mean
+  int nCl = 0;
+  float sum = 0;
+  size_t firstCl = charge.size() * low;
+  size_t lastCl = charge.size() * high;
+
+  for (size_t iCl = firstCl; iCl < lastCl; ++iCl) {
+    sum += charge[iCl];
+    ++nCl;
+  }
+
+  if (nCl > 0) {
+    sum /= nCl;
+  }
+  return sum;
+}
+
+float CalculatedEdx::getTrackTopologyCorrection(const o2::tpc::TrackTPC& track, const unsigned int region) const
+{
+  const float padLength = Mapper::instance().getPadRegionInfo(region).getPadHeight();
+  const float snp = track.getSnp();
+  const float tgl = track.getTgl();
+  const float snp2 = snp * snp;
+  const float tgl2 = tgl * tgl;
+  // calculate the trace length of the track over the pad
+  const float effectiveLength = padLength * std::sqrt((1 + tgl2) / (1 - snp2));
+  return effectiveLength;
+}
+
+float CalculatedEdx::getTrackTopologyCorrectionPol(const o2::tpc::TrackTPC& track, const o2::tpc::ClusterNative& cl, const unsigned int region, const float charge, ChargeType chargeType, const float threshold) const
+{
+  const float snp = std::abs(track.getSnp());
+  const float tgl = track.getTgl();
+  const float snp2 = snp * snp;
+  const float tgl2 = tgl * tgl;
+  const float sec2 = 1.f / (1.f - snp2);
+  const float tanTheta = std::sqrt(tgl2 * sec2);
+
+  const float z = std::abs(track.getParam(1));
+  const float padTmp = cl.getPad();
+  const float absRelPad = std::abs(padTmp - int(padTmp + 0.5f));
+  const float relTime = cl.getTime() - int(cl.getTime() + 0.5f);
+
+  const float effectiveLength = mCalibCont.getTopologyCorrection(region, chargeType, tanTheta, snp, z, absRelPad, relTime, threshold, charge);
+  return effectiveLength;
+}
+
+void CalculatedEdx::loadCalibsFromCCDB(int runNumber)
```

Comment:
It might be good to also be able to load the calibrations for a specific time stamp. One could e.g. call the parameter `long runNumberOrTimeStamp` and use the convention that if `runNumberOrTimeStamp < 10000000` it is a run number, otherwise a time stamp.